### PR TITLE
Adding getLogging functionality to s3 buckets

### PR DESCRIPTION
**Pull request by ina-stoyanova**
Adding `GetBucketLogging` function for the use of such functionality in this PR: https://github.com/gruntwork-io/terragrunt/pull/1400

### CODE REVIEW

**Comment by ina-stoyanova** [comment]
adding a test for this now...

**Comment by ina-stoyanova** [comment]
This is ready for review now. 

When this is merged & released, it will unblock us to unskip the `TestTerragruntSetsAccessLoggingForTfSTateS3BuckeToADifferentBucket` test in `terragrunt`

You can find the relevant commit with the skipped test [here](https://github.com/gruntwork-io/terragrunt/commit/426cb9462f0a1177f07cbccb9f57e41bbb91bbf4#diff-8118e44ac15fb5ab2e2ece2ab52fd23062a10b5cc9c55ee5b3620a1242cff27cR601)

File: modules/aws/s3.go
Comment (yorinasub17) [inline modules/aws/s3.go:310]: Can you share what use case prompted you to add the put version? The reason I ask is that this technically doesn't pass [the litmus test](https://terratest.gruntwork.io/docs/community/contributing/#infrastructure-and-validation-helpers) for contributions because it is an API for modifications, something that isn't really necessary in a testing library.

File: modules/aws/s3.go
Comment (yorinasub17) [inline modules/aws/s3.go:470]: This will return `nil` because `err` is `nil` at this point (it failes the first if condition, which checks for non-nil). 

That said, this is a relatively simple function on top of `GetS3BucketLoggingTarget`, so I think it fails the complexity litmus test for inclusion in terratest.

Instead, you can implement this inline in your test as:

```
targetBucket := aws.GetS3BucketLoggingTarget(...)
assert.Equal(t, targetBucket, logsTargetBucketName)
```

File: modules/aws/s3.go
Comment (yorinasub17) [inline modules/aws/s3.go:339]: I'm a bit on the fence about this one, given that it's a thin wrapper around the API, but since the s3 based functionality isn't really complete ([completeness litmus test](https://github.com/gruntwork-io/terratest/issues/421#issuecomment-567350190)) without looking for access logging, I think it is ok to include this getter function.

File: modules/aws/s3_test.go
Comment (yorinasub17) [inline modules/aws/s3_test.go:44]: If we end up removing this function, you should implement this inline. Alternatively, you should enhance https://github.com/gruntwork-io/terratest/tree/master/examples/terraform-aws-s3-example with access logging and update https://github.com/gruntwork-io/terratest/blob/master/test/terraform_aws_s3_example_test.go

**Review by yorinasub17 — COMMENTED** [review]
Implementation mostly LGTM! However, I think a few of these functions fail our [litmus test for contributions](https://terratest.gruntwork.io/docs/community/contributing/#infrastructure-and-validation-helpers).

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:310]: Yeah, that's a very good question @yorinasub17 

Initially, my thinking was "don't overclutter terratest, and just add the code I needed" _- as in the `GetLoggingTargetBucketName function`_

As I got to this yesterday though, I felt uncomfortable adding in code that was not tested, which then prompted me to add the PutBucketLogging method.

It wasn't a lot of extra effort, so I did add it in the end. Do you think it's smarter at this point to maybe skip that testing bit entirely? Or am I missing an obvious way how I can test without the prior setup? 

As it was a difficult trade-off, I thought it will be safer to push code that is tested. But as this was my main wondering _(and the reason why I didn't add the tests initially when the PR was opened)_, I'm very happy to take guidance and learn from this!

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:470]: sweet! that simplifies it a lot.

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:339]: Just checking if I've understood you correctly - is it that the function is a bit too specific?

I guess a bit of context would help - [this is actually going to help me test using `terratest` the new `terragrunt` `accesslogging_bucket_name`. ](https://github.com/gruntwork-io/terragrunt/pull/1400#pullrequestreview-518125751)

However, after you've pointed me to the `litmus test`, I can clearly see this is only needed as functionality in that one specific scenario.

File: modules/aws/s3_test.go
Comment (ina-stoyanova) [reply modules/aws/s3_test.go:44]: Yeah - if it's fine for the putting bucket logging to live inside the test only, as we don't need it for more than one scenario, I think it's better to go with it being inline! This way, we don't add extra logic into terratest, but I'm still able to test the `getLogging` function, which _was_ my original and main intention.

File: modules/aws/s3.go
Comment (yorinasub17) [reply modules/aws/s3.go:339]: The specific litmus test I am on the fence about is the "complexity" test, which states:

> Complexity: we ask that you only contribute infrastructure and validation helpers for code that is relatively complex to do from scratch. For example, a helper that merely wraps an existing function in the AWS or GCP SDK is not a great choice, as the wrapper isn’t contributing much value, but is bloating the Terratest API. On the other hand, helpers that expose simple APIs for complex logic are great contributions: ssh.CheckSshCommand is a great example of this, as it provides a simple one-line interface for dozens of lines of complicated SSH logic.

This function is a thin wrapper around the AWS SDK, which is why it can be considered as bloating the Terratest API. For example, in your test code, you can do something like the following:

```go
s3Client := NewS3Client(t, awsRegion)
res, err := GetBucketLogging(&s3.GetBucketLoggingInput{Bucket: &bucket})
require.NoError(t, err)
// Assert that logging is enabled
require.NotNil(t, res.LoggingEnabled)
// Assert that logging is done to the expected target bucket
assert.Equal(t, expectedTargetBucket, aws.StringValue(res.LoggingEnabled.TargetBucket))
```

This code is readable, and more direct in the assertions. Of course, arguably one line is better than five lines, and it will be subjective which is better (having a terratest function that simplifies these 5 lines of code, or writing them inline). Moreover, access log checks is very common, so one could argue that the terratest API is incomplete without it. This is why I am on the fence, but leaning towards including it.

File: modules/aws/s3_test.go
Comment (yorinasub17) [reply modules/aws/s3_test.go:46]: Coming back to this after a day, I actually think it might be better to do this in terraform (the alternative I suggested). That would probably be a more robust test and cleaner.

File: modules/aws/s3.go
Comment (yorinasub17) [reply modules/aws/s3.go:310]: Ah gotcha! That makes sense.

So in general, we prefer to have terraform based testing for terratest functions (integration tests) as opposed to having unit tests. There are 2 main advantages for this:

- We get to use terraform, a robust platform for managing infra, to do the test setup and teardown of resources that terratest depends on. This allows us to keep the API focused on validating infra, as opposed to managing infra.

- Users tend to refer to the examples for inspiration on testing. As such, having a terraform driven example + test, it leads to better discoverability of the functions (as opposed to a unit test which might not be as visible).

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:339]: OK, this is really insightful, Yori! Thanks!!

I've spent some time adding this as a terraform test instead. I've got a bit more cleanup to do, so will move to draft for now, and then move it back into "Ready for Review" as I work through stuff (most likely on Monday).

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:310]: this is super useful to understand - thanks for explaining! 👍

**Comment by ina-stoyanova** [comment]
I believe this is ready for another review 🙂 Moving from Draft to ready for Review.

File: examples/azure/terraform-azure-mysqldb-example/main.tf
Comment (yorinasub17) [inline examples/azure/terraform-azure-mysqldb-example/main.tf:34]: Ah sorry I caused hassle here. I just merged the fix for this in master so you might have conflicts with it.

File: examples/terraform-aws-s3-logging-enabled-example/README.md
Comment (yorinasub17) [inline examples/terraform-aws-s3-logging-enabled-example/README.md:1]: Oh I feel bad saying this after all the effort into creating a new example, but I actually meant that you should update the existing example as opposed to creating a new one (as suggested here: https://github.com/gruntwork-io/terratest/pull/689#discussion_r518288068)

This is more an art instead of a policy, but the main reason for it is that most people will look at the S3 example for inspirations on testing s3 buckets, and it would be a bit of a sprawl to have a separate example/test just for the logging test (since it is a relatively small extension of the existing s3 example).

We want to limit the number of examples/tests we add because the overhead of another example is not just UX, but it also affects our testing infra as well, as that is more code to maintain when we do, for example, a terraform version update.

Again this is a bit difficult to explain given that it is more an art of balancing tradeoffs, but in this case I think it would be favorable to add access logging to the existing s3 example and add the additional check to the existing test.

File: modules/aws/s3.go
Comment (yorinasub17) [inline modules/aws/s3.go:289]: I missed this in the first pass, but I just realized that this won't return an error, but `nil` (because you had the `err` check before). It might be more favorable to return an error in this case, since logging is not enabled.

Can you make this return a custom error (e.g., `S3AccessLoggingNotEnabledErr`)?

**Review by yorinasub17 — COMMENTED** [review]
Almost there! I had 2 more nits, but once that is addressed I think we can merge this in!

File: examples/azure/terraform-azure-mysqldb-example/main.tf
Comment (ina-stoyanova) [reply examples/azure/terraform-azure-mysqldb-example/main.tf:34]: no probs - I'll merge!

File: test/terraform_aws_s3_logging_enabled_example_test.go
Comment (ina-stoyanova) [reply test/terraform_aws_s3_logging_enabled_example_test.go:22]: hah good spot! I re-read this thing so many times and still missed it! Thanks :)

File: modules/aws/s3.go
Comment (ina-stoyanova) [reply modules/aws/s3.go:289]: Good shout!

File: examples/terraform-aws-s3-logging-enabled-example/README.md
Comment (ina-stoyanova) [reply examples/terraform-aws-s3-logging-enabled-example/README.md:1]: gotcha! no worries - it's all good experience! 

Having less code to maintain is a really good reason that resonates, and also giving people less variations will also be more straightforward! Agree :)

File: test/terraform_aws_s3_logging_enabled_example_test.go
Comment (zackproser) [reply test/terraform_aws_s3_logging_enabled_example_test.go:22]: Thanks, was also just confirming my own understanding :+1:

**Comment by ina-stoyanova** [comment]
The tests for this keep breaking for no apparent reason. When run locally, obviously all passes 🙄  I'm looking into this, but will try re-running the tests a few times and see if anything concrete comes up as a reason.

**Comment by ina-stoyanova** [comment]
Ok, now that the tests are passing, good for one more try @yorinasub17 @zackproser - would appreciate another few minutes of your time to get this merged! 🙂

File: examples/terraform-aws-s3-logging-enabled-example/README.md
Comment (yorinasub17) [reply examples/terraform-aws-s3-logging-enabled-example/README.md:1]: It looks like you still have the new example here. Was there some discussion where we decided to go with the new example instead of consolidating?

Comment (ina-stoyanova) [reply examples/terraform-aws-s3-logging-enabled-example/README.md:1]: oh no - that's a mistake! correcting it now..

Comment (ina-stoyanova) [reply examples/terraform-aws-s3-logging-enabled-example/README.md]: D'oh! I thought I pushed - but apparently hadn't done so. All pushed up now :) My bad!

File: examples/terraform-aws-s3-example/README.md
Comment (yorinasub17) [inline examples/terraform-aws-s3-example/README.md:5]: NIT:

```suggestion
Buckets](https://aws.amazon.com/s3/) - one S3 Bucket with logging and versioning enabled, and another "targetBucket" one to serve as a
```

File: examples/terraform-aws-s3-example/main.tf
Comment (yorinasub17) [inline examples/terraform-aws-s3-example/main.tf:64]: Might want to set `force_destroy = true` to ensure it can be destroyed.

File: examples/terraform-aws-s3-example/main.tf
Comment (yorinasub17) [inline examples/terraform-aws-s3-example/main.tf:56]: Our terraform style guide is to use snake case instead of kebob case for block labels.

```suggestion
resource "aws_s3_bucket" "test_bucket_logs" {
```

File: examples/terraform-aws-s3-example/outputs.tf
Comment (yorinasub17) [inline examples/terraform-aws-s3-example/outputs.tf:10]: NIT: Use `[0]` so that we return a string instead of a list:
```suggestion
  value = aws_s3_bucket.test_bucket.logging[0].target_bucket
```

**Review by yorinasub17 — COMMENTED** [review]
LGTM! Only had a few more style nits in the example, but otherwise looks good!

File: examples/terraform-aws-s3-example/main.tf
Comment (ina-stoyanova) [reply examples/terraform-aws-s3-example/main.tf:64]: good point!

File: examples/terraform-aws-s3-example/outputs.tf
Comment (ina-stoyanova) [reply examples/terraform-aws-s3-example/outputs.tf:10]: This is what i tried initially, but then came across [this issue:](https://github.com/hashicorp/terraform/issues/22504)
```
Block type "logging" is represented by a set of objects, and set elements do
        	            	not have addressable keys. To find elements matching specific criteria, use a
        	            	"for" expression with an "if" clause.
```

I thought the simplest way is leave this as is. We have a few other ways we could go about this to make it clearer though:

- add a comment above this line
- use `tolist(aws_s3_bucket.test_bucket.logging)[0].target_bucket` (I've tested this and it seemed to work well!)
- or try changing the lifecycle policy for this: _(as suggested in some of the issue threads)_
```
lifecycle {
     create_before_destroy = true
     ignore_changes = [
       # Ignore changes to first tag in list: "foo"
       tags[0]
     ]
   }
```

I'd probably go with the 2nd option - use the `tolist` tf function, so it's clear to users of this example this returns a list

Comment (yorinasub17) [reply examples/terraform-aws-s3-example/outputs.tf]: Ah gotcha. Yes the `tolist` approach works for me!

**Review by yorinasub17 — APPROVED** [review]
Updates LGTM!

**Comment by ina-stoyanova** [comment]
Sweet! Thanks for the help and the patience @yorinasub17! 🎉

**Pull request merged**
